File: app/src/main/java/com/yatatsu/passiveviewsample/ui/base/Controller.java
package com.yatatsu.passiveviewsample.ui.base;


import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

public abstract class Controller<T extends Screen> {

    private T screen;
    private CompositeSubscription compositeSubscription;

    void registerScreen(T screen) {
        this.screen = screen;
    }

    protected T getScreen() {
        return screen;
    }

    protected void register(Subscription subscription) {
        if (compositeSubscription == null) {
            compositeSubscription = new CompositeSubscription();
        }
        compositeSubscription.add(subscription);
    }

    protected void unSubscribe() {
        if (compositeSubscription != null) {
            compositeSubscription.unsubscribe();
            compositeSubscription = null;
        }
    }
}

File: app/src/main/java/com/yatatsu/passiveviewsample/ui/base/Controller.kt
package com.yatatsu.passiveviewsample.ui.base


import rx.Subscription
import rx.subscriptions.CompositeSubscription

abstract class Controller<T : Screen> {

  protected var screen: T? = null
    private set
  private var compositeSubscription: CompositeSubscription? = null

  fun registerScreen(screen: T) {
    this.screen = screen
  }

  protected fun register(subscription: Subscription) {
    if (compositeSubscription == null) {
      compositeSubscription = CompositeSubscription()
    }
    compositeSubscription!!.add(subscription)
  }

  protected fun unSubscribe() {
    if (compositeSubscription != null) {
      compositeSubscription!!.unsubscribe()
      compositeSubscription = null
    }
  }
}
